# 899 Omit overrideSearchType, when knowledge base resource contains …

ShotaOki opened the pull request:
*Issue #, if available:*

#899 [BUG] When Kendra is set up in the knowledge base, search processing hangs due to an error.

*Description of changes:*

Implemented the following actions:

1. When saving the bot, execute BedrockAgent::get_knowledge_base to retrieve the type of knowledge base.
Store the type in the searchParam of DynamoDB.

2. Before executing agent_client.retrieve, retrieve the type of knowledge base from the searchParam of DynamoDB.
If the knowledge base type is KENDRA, remove the overrideSearchType parameter and re-execute.

3. If there is an error during the execution of retrieve, determine whether the error message contains Kendra.
If Kendra is included, remove the overrideSearchType parameter and re-execute.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ShotaOki committed:
# 899 Omit overrideSearchType, when knowledge base resource contains Kendra
----
If the backend data source type (i.e. Opensearch, Kendra, etc) can be determined before invoke agent_client.retrieve, omit overrideSearchType option when the type is KENDRA
If not possible, catch exception and analyze message whether contain Kendra or not, and handle the case

ShotaOki committed:
fixed mypy check errror.

ShotaOki committed:
fixed black formatted

Yukinobu-Mine commented on backend/app/vector_search.py:
```diff
@@ -83,16 +83,52 @@ def _bedrock_knowledge_base_search(bot: BotModel, query: str) -> list[SearchResu
+        if (
+            bot.bedrock_knowledge_base.search_params.knowledge_base_resource_type
+            == "KENDRA"
+        ):
```
この部分で `get_knowledge_base_info()` を実行するようにすれば、`SearchParamsModel` や `SearchParams` に `knowledge_base_resource_type` を追加する必要は無いのではないかと思うのですが、いかがでしょうか?

ShotaOki replied:
ありがとうございます！ こちらいただいたご指摘の内容で対応いたしました

Yukinobu-Mine commented on backend/app/repositories/custom_bot.py:
```diff
@@ -95,6 +96,25 @@ def store_bot(custom_bot: BotModel):
+        # Set Knowledge Base Resource Type
+        custom_bot.bedrock_knowledge_base.search_params.knowledge_base_resource_type = (
+            knowledge_base_type
+        )
```
`vector_search.py` 内で `get_knowledge_base_info()` を実行するようにすれば、このような処理は不要になるのではないかと思ったのですが、いかがでしょうか?

ShotaOki replied:
ありがとうございます！ こちらいただいたご指摘の内容で対応いたしました

Yukinobu-Mine commented on backend/app/vector_search.py:
```diff
@@ -83,16 +83,52 @@ def _bedrock_knowledge_base_search(bot: BotModel, query: str) -> list[SearchResu
-            retrievalQuery={"text": query},
-            retrievalConfiguration={
+        # Init retrieve parameter
+        retrieve_parameter = {
```
`retrieve_parameter: KnowledgeBaseRetrievalConfigurationTypeDef = {` とし、12行目の import に `KnowledgeBaseRetrievalConfigurationTypeDef` を追加することで type safety を確保する事ができますので、ご検討頂けますと幸いです。

ShotaOki replied:
ありがとうございます！ こちらretrieve関数の引数の型をmypyから参照して、type safetyを確保するようにいたしました

Yukinobu-Mine commented on backend/app/vector_search.py:
```diff
@@ -83,16 +83,52 @@ def _bedrock_knowledge_base_search(bot: BotModel, query: str) -> list[SearchResu
+        try:
+            # Send retrieve request
+            response = agent_client.retrieve(**retrieve_parameter)
+        except ClientError as e:
```
このファイルの中で事前に `get_knowledge_base_info()` を実行するようにすれば、この例外処理も不要になるのではないかと思うのですが、いかがでしょうか?

ShotaOki replied:
ありがとうございます！ こちらいただいたご指摘の内容で対応いたしました

Yukinobu-Mine commented:
@ShotaOki いくつかコメントさせて頂きました。ご確認頂けますと幸いです。

ShotaOki commented:
ありがとうございます！
こちら対応いたします！！

ShotaOki committed:
# 899 Omit overrideSearchType, when knowledge base resource contains
- remove dynamoDB cache
- move get_knowledge_base_info to vector_search.py
- add type annotation

Yukinobu-Mine commented on backend/app/repositories/models/custom_bot_kb.py:
```diff
@@ -60,6 +61,18 @@ class WebCrawlingFiltersModel(BaseModel):
 
 
+class KnowledgeBaseConfiguration(BaseModel):
+    type: type_kb_resource_type = "VECTOR"
```
このデフォルト値 `VECTOR` は不要だと思うので、デフォルト値の指定を削除して頂くことは可能でしょうか?

ShotaOki replied:
ありがとうございます！ こちらいただいたご指摘の内容で対応いたしました

Yukinobu-Mine commented:
@ShotaOki 対応ありがとうございます。1点だけ追加でコメントさせて頂いたので、ご対応頂けますと幸いです。

ShotaOki committed:
# 899 Omit overrideSearchType, when knowledge base resource contains
- Remove type_kb_resource_type default value

ShotaOki commented:
@Yukinobu-Mine こちらいただいた指摘について対応しました！ また、ローカルで正常に動作することを確認しました。こちらよろしくお願いいたします！

Yukinobu-Mine commented:
@ShotaOki マージしました。コントリビュートありがとうございます！
